src/websocket_stream.cpp

```diff
@@ -0,0 +1,308 @@
+
+	do_tcp_connect(std::move(endpoints));
+}
+
+void websocket_stream::do_tcp_connect(std::vector<tcp::endpoint> endpoints)
+{
+	m_endpoints = std::move(endpoints);
+
+	auto tcp_stream = std::visit(overloaded
+		{
+			[&](stream_type &stream) { return &stream.next_layer(); }
+			, [&](ssl_stream_type &stream) { return &stream.next_layer().next_layer(); }
+		}
+		, m_stream);
+
+	ADD_OUTSTANDING_ASYNC("websocket_stream::on_tcp_connect");
+	boost::asio::async_connect(*tcp_stream
+		, m_endpoints.rbegin()
+		, m_endpoints.rend()
+		, std::bind(&websocket_stream::on_tcp_connect, shared_from_this(), _1));
+}
+
+void websocket_stream::on_tcp_connect(error_code const& ec)
+{
+	COMPLETE_ASYNC("websocket_stream::on_tcp_connect");
+	if (ec)
+    {
+		if (auto handler = std::exchange(m_connect_handler, nullptr))
+			post(m_io_service, std::bind(handler, ec));
+        return;
+    }
+
+	if (std::holds_alternative<ssl_stream_type>(m_stream)) do_ssl_handshake();
+	else do_handshake();
+}
+
+void websocket_stream::do_ssl_handshake()
+{
+	auto& ssl_stream = std::get<ssl_stream_type>(m_stream).next_layer();
+
+	error_code ec;
+	ssl::set_host_name(ssl::get_handle(ssl_stream), m_hostname, ec);
+    if (ec)
+    {
+		if (auto handler = std::exchange(m_connect_handler, nullptr))
+			post(m_io_service, std::bind(handler, ec));
+        return;
+    }
+
+	ADD_OUTSTANDING_ASYNC("websocket_stream::on_ssl_handshake");
+	ssl_stream.async_handshake(ssl::stream_base::client
+			, std::bind(&websocket_stream::on_ssl_handshake, shared_from_this(), _1));
+}
+
+void websocket_stream::on_ssl_handshake(error_code const& ec)
+{
+	COMPLETE_ASYNC("websocket_stream::on_ssl_handshake");
+	if (ec)
+	{
+		if (auto handler = std::exchange(m_connect_handler, nullptr))
+			post(m_io_service, std::bind(handler, ec));
+		return;
+	}
+
+	do_handshake();
+}
+
+void websocket_stream::do_handshake()
+{
+	auto user_agent_handler = [user_agent = m_user_agent](websocket::request_type& req)
+		{
+			if (!user_agent.empty()) req.set(http::field::user_agent, user_agent);
+		};
+
+	std::string host = m_hostname;
+	if(m_port != 80 && m_port != 443)
+		host += std::string(":") + std::to_string(m_port);
+
+	ADD_OUTSTANDING_ASYNC("websocket_stream::on_handshake");
+	std::visit([&](auto& stream)
+		{
+#if BOOST_VERSION >= 107000
+			stream.set_option(websocket::stream_base::decorator(user_agent_handler));
+			stream.async_handshake(host
+				, m_target
+				, std::bind(&websocket_stream::on_handshake, shared_from_this(), _1));
+#else
+			stream.async_handshake_ex(host
+				, m_target
+				, user_agent_handler
+				, std::bind(&websocket_stream::on_handshake, shared_from_this(), _1));
+#endif
+	}
+	, m_stream);
+}
+
+void websocket_stream::on_handshake(error_code const& ec)
+{
+	COMPLETE_ASYNC("websocket_stream::on_handshake");
+	auto handler = std::exchange(m_connect_handler, nullptr);
+	if (ec)
+    {
+		if (handler) post(m_io_service, std::bind(handler, ec));
+        return;
```

Comment:
I really should get a clang-format file... :)

Reply:
Fixed!